### PR TITLE
[Plutus Core] [Builtins] Added the 'sum' and 'product' dynamic builtin types

### PR DESCRIPTION
**Pull request by effectfully**
This adds the `sum` and `product` dynamic builtin types. Turned out to be trickier that I thought, because in order to emit elements out of sums and products we need two distinct emitters and the existing API did not support that properly. Unfortunately, I'm not able to find a nice API, so I just replicated the dynamic builtin handlers approach from `Language.PlutusCore.Constant.Dynamic.OnChain`, but specified it to be only about emitters. It's not terrible, I believe, just a bit wordy: we need to do the following:

```haskell
unsafePerformIO . withEmitHandler eval $
    withEmitTerm TypedBuiltinDyn $ \emitX ->
    withEmitTerm TypedBuiltinDyn $ \emitY ->
        feedEmitHandler $ go emitX emitY
```

in order to be able to use two emitters from within `go` (also I now realize that those `TypedBuiltinDyn` arguments can be inferred, but that's for another PR).

### CODE REVIEW

File: language-plutus-core/generators/Language/PlutusCore/Generators/Interesting.hs [hunk 1]
```diff
@@ -181,6 +183,31 @@ genNatRoundtrip = do
         return $ mkIterApp () (TyInst () natToInteger size) [ssize, n]
     return . TermOf term $ TypedBuiltinValue typedIntS iv
 
+-- | @sumNat@ as a PLC term.
+getBuiltinNatSum :: Size -> Quote (Term TyName Name ())
+getBuiltinNatSum s = rename =<< do
```
Comment (effectfully) [inline language-plutus-core/generators/Language/PlutusCore/Generators/Interesting.hs]: Moved this from `Meta.hs` as it's not a standard library combinator, just some weird function that turns a list of natural into a list of integers and adds them up. It also receives a `Size`, because when it was written there were no handy combinators to deal with sized types, so this should be fixed. later

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Emit.hs [hunk 2]
```diff
@@ -32,19 +38,35 @@ globalUniqueVar = unsafePerformIO $ newIORef 0
 {-# NOINLINE globalUniqueVar #-}
 
 nextGlobalUnique :: IO Int
-nextGlobalUnique = atomicModifyIORef' globalUniqueVar $ \i -> (i, succ i)
+nextGlobalUnique = atomicModifyIORef' globalUniqueVar $ \i -> (succ i, i)
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Emit.hs]: That was an amazingly stupid bug.

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+import           Language.PlutusCore.MkPlc
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Renamer
+import           Language.PlutusCore.Type
+
+-- | '(,)' as a PLC type.
+--
+-- > /\(a b :: *) -> all (r :: *). (a -> b -> r) -> r
```
Comment (michaelpj) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs]: Should be `\` not `/\`

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Sum.hs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    ) where
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Renamer
+import           Language.PlutusCore.Type
+
+-- | 'Either' as a PLC type.
+--
+-- > /\(a b :: *) -> all (r :: *). (a -> r) -> (b -> r) -> r
```
Comment (michaelpj) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Sum.hs]: `\`

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Unit.hs [hunk 5]
```diff
@@ -34,3 +35,17 @@ getBuiltinUnitval = rename =<< do
         . TyAbs () a (Type ())
         . LamAbs () x (TyVar () a)
         $ Var () x
+
+-- | @seq @() @()@ as a PLC term.
+--
+-- > \(x y : unit) -> unitval
+getBuiltinSequ :: Quote (Value TyName Name ())
```
Comment (michaelpj) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Unit.hs]: I don't understand this - we're strict so why do we have this `seq` thing?

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs [hunk 6]
```diff
@@ -51,33 +50,109 @@ instance KnownDynamicBuiltinType Char where
...
+        foldl1 (TyApp ()) <$> sequence
+            [ getBuiltinProduct
+            , getTypeEncoding $ Proxy @a
+            , getTypeEncoding $ Proxy @b
+            ]
+
+    makeDynamicBuiltin (x, y) = do
+        da <- getTypeEncoding $ Proxy @a
+        db <- getTypeEncoding $ Proxy @b
+        -- TODO: perhaps we should use 'MaybeT' in the type signature of 'makeDynamicBuiltin'.
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: :+1:

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Unit.hs (hunk 5 above)
Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Unit.hs]: `emit` on the PLC side has this type: `emit : a -> ()` for some particular `a`. Now if we want to emit two values, we write `sequ (emit x) (emit y)`. So yes, we're strict, but we're not pure and that's why `sequ` exists.

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs [hunk 7]
```diff
@@ -51,33 +50,109 @@ instance KnownDynamicBuiltinType Char where
 
 instance PrettyDynamic Char
 
+instance (KnownDynamicBuiltinType a, KnownDynamicBuiltinType b) =>
+            KnownDynamicBuiltinType (a, b) where
+    getTypeEncoding _ =
+        foldl1 (TyApp ()) <$> sequence
```
[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: There is `foldl1` there and that's why I didn't write `mkIterTyApp`, but you're right, a bit of massaging and we can use `mkIterTyApp`.

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs (hunk 3 above)
Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs]: Thanks.

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs (hunk 6 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: Type checker refactoring has higher priority, though.

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+getBuiltinProduct :: Quote (Type TyName ())
+getBuiltinProduct = rename =<< do
```
[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs]: Right, fixing.

[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Product.hs]: Done. I used tuples instead of products in the definition of the `KnownDynamicBuiltinType (a, b)` instance.

**Comment by effectfully** [comment]
Comments addressed.
